crates/services/txpool_v2/src/storage/graph.rs

```diff
@@ -0,0 +1,477 @@
+            .collect();
+        let Some(root) = self.graph.remove_node(root_id) else {
+            return Ok(vec![]);
+        };
+        let mut removed_transactions = vec![root.transaction];
+        for dependent in dependents {
+            removed_transactions.extend(self.remove_dependent_sub_graph(dependent)?);
+        }
+        Ok(removed_transactions)
+    }
+
+    /// Check if the input has the right data to spend the output present in pool.
+    fn check_if_coin_input_can_spend_output(
+        output: &Output,
+        input: &Input,
+    ) -> Result<(), Error> {
+        if let Input::CoinSigned(CoinSigned {
+            owner,
+            amount,
+            asset_id,
+            ..
+        })
+        | Input::CoinPredicate(CoinPredicate {
+            owner,
+            amount,
+            asset_id,
+            ..
+        }) = input
+        {
+            let i_owner = owner;
+            let i_amount = amount;
+            let i_asset_id = asset_id;
+            match output {
+                Output::Coin {
+                    to,
+                    amount,
+                    asset_id,
+                } => {
+                    if to != i_owner {
+                        return Err(Error::NotInsertedIoWrongOwner);
+                    }
+                    if amount != i_amount {
+                        return Err(Error::NotInsertedIoWrongAmount);
+                    }
+                    if asset_id != i_asset_id {
+                        return Err(Error::NotInsertedIoWrongAssetId);
+                    }
+                }
+                Output::Contract(_) => return Err(Error::NotInsertedIoContractOutput),
+                Output::Change { .. } => {
+                    return Err(Error::NotInsertedInputDependentOnChangeOrVariable)
+                }
+                Output::Variable { .. } => {
+                    return Err(Error::NotInsertedInputDependentOnChangeOrVariable)
+                }
+                Output::ContractCreated { .. } => {
+                    return Err(Error::NotInsertedIoContractOutput)
+                }
+            };
+        }
+        Ok(())
+    }
+
+    /// Cache the transaction information in the storage caches.
+    /// This is used to speed up the verification/dependencies searches of the transactions.
+    fn cache_tx_infos(
+        &mut self,
+        outputs: &[Output],
+        tx_id: &TxId,
+        node_id: NodeIndex,
+    ) -> Result<(), Error> {
+        for (index, output) in outputs.iter().enumerate() {
+            let index = u16::try_from(index).map_err(|_| {
+                Error::WrongOutputNumber(format!(
+                    "The number of outputs in `{}` is more than `u8::max`",
+                    tx_id
+                ))
+            })?;
+            let utxo_id = UtxoId::new(*tx_id, index);
+            match output {
+                Output::Coin { .. } | Output::Change { .. } | Output::Variable { .. } => {
```

Comment:
We shouldn't allow dependencies over `Change` and `Variable` outputs since it is impossible to verify the owner, amount, or asset id. It looks like in another place, you will return a `NotInsertedInputDependentOnChangeOrVariable` error if someone tries to use it as an input. Maybe it is better just not to add this kind of coins in `self.coins_creators` at all?

Reply:
Yeah you are right it's more safe. Done.